onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1882,37 +1920,11 @@
             // keep node index first, will turn it to pc index later
             execution_plan[i]->step_pc.push_back(node_index);
 #endif
-          }
-          // check whether we need to wait on notification
-          auto notification_it = node_to_notification.find(it->Index());
-          if (notification_it != node_to_notification.end()) {
-            // push a wait command if has EP registered it.
-            // find which tensor we consumed to decide which wait handle to use
-            // if there are multiple tensors consumed on different devices, need to wait on all of them
-            // for example, if a cuda kernel A produce a cpu tensor C and a gpu tensor C', cuda kernel B
-            // depends on both C and C', we need to wait on both GPU and CPU.
-            InlinedHashSet<OrtDevice::DeviceType> devices_to_wait_on;
-            for (auto* inputs : node->InputDefs()) {
-              if (inputs->Exists()) {
-                if (std::find(it->OutputDefs().begin(), it->OutputDefs().end(), inputs) != it->OutputDefs().end()) {
-                  OrtValueIndex input_arg_idx;
-                  ORT_THROW_IF_ERROR(ort_value_name_idx_map_.GetIdx(inputs->Name(), input_arg_idx));
-                  auto& consumer_device = plan_.allocation_plan[input_arg_idx].location.device;
-                  if (devices_to_wait_on.find(consumer_device.Type()) == devices_to_wait_on.end()) {
-                    auto wait_handle = stream_handle_registry.GetWaitHandle(
-                        execution_plan[plan_.notification_owners[notification_it->second]]->device_.Type(),
-                        consumer_device.Type());
-                    if (wait_handle) {
-                      execution_plan[i]->steps_.emplace_back(std::make_unique<WaitOnEPStep>(wait_handle, notification_it->second));
-#ifdef ENABLE_TRAINING
-                      execution_plan[i]->step_pc.push_back(node_index);
-#endif
-                    }
-                    devices_to_wait_on.insert(consumer_device.Type());
-                  }
-                }
-              }
-            }
+            ConstructWaitOnEPStep(node, *it, trigger_point_it->second, stream_handle_registry, execution_plan, i, devices_to_wait_on);
+          } else if ((notification_it = node_to_notification.find(it->Index())) != node_to_notification.end()) {
+            // Else the current node and its input node are in the same stream, but we can still find input node in node_to_notification,
+            // which means we will launch WaitOnEPStep only before LaunchKernelStep.
```

Comment:
For "else" what if we do not launch WaitOnEpStep? Still we have the right sequence no?

Reply:
If we do not launch WaitOnEpStep under "else if" clause, that means there exist a consumer node which is in the same stream as its producer node, and it should have launched WaitOnEpStep (for the corresponding tensors are in different device). We will not have the right sequence for this case